contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+import "./core/Safe.sol";
+import "@gnosis.pm/safe-contracts/contracts/proxies/GnosisSafeProxyFactory.sol";
+import "./libraries/SafeERC677.sol";
+
+import "hardhat/console.sol";
+
+contract PrepaidCardMarketV2 is Ownable, Versionable, Safe {
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+  using SafeERC20Upgradeable for IERC677;
+  using SafeERC677 for IERC677;
+
+  struct SKU {
+    address issuerSafe;
+    address issuer;
+    address issuingToken;
+    uint256 faceValue;
+    string customizationDID;
+  }
+
+  address public prepaidCardManagerAddress;
+  address public tokenManager;
+  address public provisioner;
+  address public versionManager;
+  address public exchangeAddress;
+  mapping(address => mapping(address => uint256)) public balance; // issuer safe address -> token -> balance
+  mapping(address => address) public issuers; // issuer safe address -> issuer EOA
+  mapping(bytes32 => SKU) public skus; // sku => sku data
+  mapping(bytes32 => uint256) public asks; // sku => ask price (in issuing token)
+
+  bool public paused;
+
+  EnumerableSetUpgradeable.AddressSet internal trustedProvisioners;
+
+  event Setup();
+
+  event TokensDeposited(
+    address issuer,
+    uint256 amount,
+    address token,
+    address safe
+  );
+  event TokensWithdrawn(
+    address safe,
+    address issuer,
+    address token,
+    uint256 amount
+  );
+  event SkuAdded(
+    address issuer,
+    address issuingToken,
+    uint256 faceValue,
+    string customizationDID,
+    bytes32 sku
+  );
+  event AskSet(
+    address issuer,
+    address issuingToken,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event ProvisionedPrepaidCard(
+    address prepaidCard,
+    address customer,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event TrustedProvisionerAdded(address token);
+  event TrustedProvisionerRemoved(address token);
+  event PausedToggled(bool paused);
+
+  function initialize(address owner) public override initializer {
+    paused = false;
+    OwnableInitialize(owner);
+  }
+
+  function setup(
+    address _exchangeAddress,
+    address _prepaidCardManagerAddress,
+    address _provisioner,
+    address _tokenManager,
+    address[] calldata _trustedProvisioners,
+    address _versionManager
+  ) external onlyOwner {
+    exchangeAddress = _exchangeAddress;
+    prepaidCardManagerAddress = _prepaidCardManagerAddress;
+    provisioner = _provisioner;
+    tokenManager = _tokenManager;
+    versionManager = _versionManager;
+
+    for (uint256 i = 0; i < _trustedProvisioners.length; i++) {
+      _addTrustedProvisioner(_trustedProvisioners[i]);
+    }
+
+    emit Setup();
+  }
+
+  function setPaused(bool _paused) external onlyOwner {
+    paused = _paused;
+    emit PausedToggled(_paused);
+  }
+
+  function getTrustedProvisioners() external view returns (address[] memory) {
+    return trustedProvisioners.values();
+  }
+
+  function _addTrustedProvisioner(address _token) internal {
```

Comment:
let's add an `onlyOwner` to this to be defensive--it's very good for documentation as well.